src/telemetry.ts

```diff
@@ -25,17 +26,59 @@ export interface ITelemetryReporter {
 
 export class TelemetryReporter implements ITelemetryReporter {
     private _sendEvent: (event: DebugProtocol.Event) => void;
+    private _globalTelemetryProperties: any = {};
 
     reportEvent(name: string, data?: any): void {
         if (this._sendEvent) {
-            const event = new OutputEvent(name, 'telemetry', data);
+            const combinedData = Object.assign({}, this._globalTelemetryProperties, data);
+            const event = new OutputEvent(name, 'telemetry', combinedData);
             this._sendEvent(event);
         }
     }
 
     setupEventHandler(_sendEvent: (event: DebugProtocol.Event) => void): void {
         this._sendEvent = _sendEvent;
     }
+
+    public addCustomGlobalProperty(additionalGlobalTelemetryProperties: any): void {
+        Object.assign(this._globalTelemetryProperties, additionalGlobalTelemetryProperties);
+    }
+}
+
+// If you add an async global property, all events after that will include it
+export class AsyncGlobalPropertiesTelemetryReporter implements ITelemetryReporter {
+    private _actionsQueue = Promise.resolve();
+
+    constructor(private _telemetryReporter: TelemetryReporter) {
+        // We just store the parameter
+    }
+
+    public reportEvent(name: string, data?: any): void {
+        this._actionsQueue = this._actionsQueue.then(() => // We block the report event until all the addCustomGlobalProperty have finished
+            this._telemetryReporter.reportEvent(name, data));
+    }
+
+    public setupEventHandler(_sendEvent: (event: DebugProtocol.Event) => void): void {
+        this._telemetryReporter.setupEventHandler(_sendEvent);
+    }
+
+    public addCustomGlobalProperty(additionalGlobalPropertiesPromise: Promise<any> | any): void {
+        this._actionsQueue = this._actionsQueue.then(async () => {
```

Comment:
Here's a suggestion for a different implementation. It might get the intent across a little better since yours is acting like a queue here but doesn't really need to be. But it's not a big difference so I leave it up to you which one to use.

```ts
const reportedPropertyP = additionalGlobalPropertiesPromise.then(prop => this._telemetryReporter.addCustomGlobalProperty, err => this.reportErrorWhileWaitingForProperty(err);

this._actionsQueue = Promise.join(
	this._actionsQueue,
	reportedPropertyP)
```

Reply:
I'm using this suggestion with Promise.all

Reply:
Ah that's what I meant.